aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+
+#include <ATen/FunctionalTensorWrapper.h>
+
+#include <ATen/FunctionalInverses.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+#include <c10/util/Exception.h>
+
+#include <c10/util/irange.h>
+
+namespace at {
+
+FunctionalTensorWrapper::FunctionalTensorWrapper(Tensor value)
+  : c10::TensorImpl(
+      c10::Storage(c10::make_intrusive<functionalization::FunctionalStorageImpl>(value.device(), value.numel(), value.dtype())),
+      c10::DispatchKeySet(DispatchKey::Functionalize),
+      value.dtype()
+    ),
+    value_(value),
+    // Note: "level" is a concept that we don't know how to compute in core.
+    // For now I'm retroactively setting this in functorch,
+    // but once Open Multiple Dispatch lands we should be able to calculate this in core.
+    level_(-1)
+{
+  // shallow_copy_from overwrites the storage and dispatch keyset...
+  auto functional_storage = storage_;
+  shallow_copy_from(value.getIntrusivePtr());
+  storage_ = functional_storage;
+  // TODO: should FunctionalTensorWrapper set_storage_access_should_throw()?
+  // Their storage is similar to meta tensors, which does not throw.
+  // However, users might expect to legally use storage() inside of a functionalized program, and this could be a foot gun.
+  storage_access_should_throw_ = false;
```

Comment:
I think false is the correct setting here, for precisely the reason given

Reply:
The thing I was worried about is, won't users silently get garbage if they try to access `tensor.storage()` from python (or is that something that people ever really do)?

For meta tensors that's probably okay (once we make storage accessible), since as a user you know that you have a meta tensor.

But in a given function defined in python, wrapping your function in a `functionalize()` will turn all intermediate tensors into "functional" tensors that don't have storage.

Either way, I kinda need `storage_acess_should_throw_=false` out of convenience in order to be able to take advantage of the `at::native::{view}` kernels to propagate strides.

Reply:
This ought to help: https://github.com/pytorch/pytorch/pull/65340/files#diff-da1fb649c7f98b7dc2910d03e27696c3e413fd248ad40468b750c8cc6092e993 It will prevent getting the storage from Python when it is nullptr